### ISSUE TITLE
Updating intellisense everytime i switch file

### ISSUE THREAD

**Issue by psclkhoury** [microsoft/vscode-cpptools#3523]
Hi,

We have a big repo in c++ and we use vs code to open it. We configured the c_cpp_properties.json to add the correct include paths and defines.

The problem happens when I switch between files:

I open Class1.cpp, i see the red flame at the bottom right that intellisense is updating. It takes like 30 seconds to a minute to finish (which is a bit long to begin with imo). Now I open Class2.cpp without closing Class1.cpp. I again get the intellisense is updating message.

The problem is when I switch back to Class1.cpp, I again get the same red flame, and it takes the same amount of time for intellisense to update. Switching again to Class2.cpp and again intellisense is updating.

This makes it unusable because I need to wait for intellisense to finish updating before I get any sort of autocomplete or type information and this takes half a minute on average everytime I switch files.
 
I tried to increase the intellisense max chache size but this didnt make any difference. Now the cache is at 30BG and intellisense is only using 10GB.

**Comment by sean-mcmanus** [comment]
We need more info to help narrow down the cause.
1. What is the lifetime of the Microsoft.VSCode.CPP.IntelliSense proceses when you do your repro? Under normal conditions, you should see 1 process open for each .cpp file and they process should not close/restart.
2. What is the memory usage for the processes? We have some code that will close unused processes after 1.5 GB of memory is used.
3. Approximately how large are the files? We have a known issue with updating for really large files (100k+ lines), but I haven't seen 30 seconds to update for normal size files yet.
4. What is your OS?
5. Does setting your intelliSenseCacheSize to 0 or back to 5120 have any effect?
6. What version of the extension are you using?

**Comment by psclkhoury** [comment]
@sean-mcmanus 

1 & 2. I didn't know that each cpp file will have its own process. I checked in the task manager and I could see that for each cpp file that I open a new C/C++ extension process was spawned.
However the memory usage (of vscode in general) was quickly going over 1.5 GB and some processes were starting to get closed.

In the picture below I had 3 tabs open and intellisense was still updating.
<img width="610" alt="vs_code_2" src="https://user-images.githubusercontent.com/9846892/56717750-a1feb100-673d-11e9-8886-7b0867f5ec1f.png">

3. the files are not large at all, most are in the order of ~100 to ~200 KB. But I always need at least 30 seconds before intellisense kicks in, could it be because our c_cpp_properties.json contains many include paths (when I say many I mean a bit more than 1000)? 
In visual studio 2019, a newly opened file takes 2 to 4 seconds before intellisense kicks in, but there we are using CppProperties.json which have a much smaller list of include paths for each subfolder, which probably makes it easier.

4. I am using windows 10

5. I deleted all the cache yesterday and I noticed that after I open the file once, even if the process is closed because of high memory usage, when I switch back to the file intellisense takes 2 seconds to kick in, so that's good. The only problem is that the first time I need to wait a bit long to get intellisense.

6. I am using Version 0.23.0-insiders: April 23, 2019 but I noticed the same problems with all the previous versions that I used.

Thank you @sean-mcmanus,  you are always very quick to respond to issues!

**Comment by sean-mcmanus** [comment]
The memory usage limit only counts for the Microsoft.VSCode.CPP.IntelliSense.Msvc.exe processes and not all of the VS Code processes and it will cause us to close older processes to free up memory, but that does not appear to be the problem you are hitting.

Having 1000 include paths is probably the cause of the 30 second slowness -- are those paths obtained recursively via paths ending with "**" or just 1000 normal paths?

You can use the C/C++: Log Diagnostics command to see the includePaths we end up sending to the IntelliSense process -- if too many includePaths are sent, it can slow down parsing considerably.

The IntelliSense cache is designed to allow skipping over parsing of header files.

**Comment by psclkhoury** [comment]
@sean-mcmanus we tried with with paths ending in ** but intellisense didn't find all the include paths so now we have 1000 normal paths.

They are all in a mapped drive called 'u:/', with 'u:/**' not all of them were found.

**Comment by sean-mcmanus** [comment]
1. I'm able to repro an issue where switching between large files causes the Microsoft.VSCode.Cpp.IntelliSense.Msvc updating to take longer and use more memory after every switch, so that might be what you are seeing, but using 1000k include paths didn't have any effect (this may be a duplicate of a known issue we've already started investigating, not sure yet).
2. I'm also reproing a 2nd issue with the 1000k include paths if I also have 1000 #include statements that causes the Microsoft.VSCode.CPP.Extension process to take a really long time, but only after the 1st open (afterwards issue 1 starts to be hit).

Which process is causing the problem you are experiencing?

**Comment by sean-mcmanus** [comment]
One potential cause of the slowness is if the IntelliSense engine has to traverse the 1000 includePaths repeatedly due to the common paths being near the end of the list (I can double check that on Monday). The recursive includes case optimizes this via removing the unused paths and moving the commonly used paths to the top of the list. We could potentially do the unused path removal if there are "too many" manually specified inludePaths.

**Comment by sean-mcmanus** [comment]
Also, we have a known issue with the IntelliSense path caches getting thrashed on Windows when in non-WSL mode due to different casings of the same file being used. I can check Monday if I'm able to detect any performance difference when using "/usr/bin/g++" instead of cl.exe.

**Comment by psclkhoury** [comment]
@sean-mcmanus I finally copied all the include files to another location and reorganized them so that they can be resolve in around 150 include paths and I see some speed improvements (intellisense needs 5 to 35 seconds to update, depending of the file). I can probably organize them even more to further reduce the number of include paths. 

I also made all the include paths lower case to avoid casing issues.

thanks
